content/library/advanced-features/app-testing/get-started.md

```diff
@@ -0,0 +1,267 @@
+
+The returned sequence of elements are ordered by appearance on the page. If containers are used to insert elements in a different order, these sequences may not match the order they are executed in your code. Consider the following example where containers are used switch the order of two buttons on the page:
+
+```python
+import streamlit as st
+
+first = st.container()
+second = st.container()
+
+second.button("A")
+first.button("B")
+```
+
+If the above app was tested, the first button (`at.button[0]`) would be labeled "B" and the second button (`at.button[1]`) would be labeled "A". As true assertions, these would be:
+
+```python
+assert at.button[0].label == "B"
+assert at.button[1].label == "A"
+```
+
+#### Retrieve widgets by key
+
+You can retrieve keyed widgets by their keys instead of their order on the page. The key of the widget is passed as either an arg or kwarg. For example, look at this app and the following (true) assertions:
+
+```python
+import streamlit as st
+
+st.button("Next", key="submit")
+st.button("Back", key="cancel")
+```
+
+```python
+assert at.button(key="submit").label == "Next"
+assert at.button("cancel").label == "Back"
+```
+
+#### Retrieve containers
+
+You can also narrow down your sequences of elements by retrieving specific containers. Each retrieved container has the same attributes as `AppTest`. For example, `at.sidebar.checkbox` returns a sequence of all checkboxes in the sidebar. `at.main.selectbox` returns the sequence of all selectboxes in the main body of the app (not in the sidebar).
+
+For `AppTest.columns` and `AppTest.tabs`, a sequence of containers is returned. So `at.columns[0].button` would be the sequence of all buttons in the first column appearing in the app.
+
+### How to manipulate widgets
+
+All widgets have a universal `.set_value()` method. Additionally, many widgets have specific methods for manipulating their value. The names of [Testing element classes](/library/api-reference/app-testing/testing-element-classes) closely match the names of the `AppTest` attributes. For example, look at the return type of [`AppTest.button`](/library/api-reference/app-testing/st.testing.v1.apptest#apptestbutton) to see the corresponding class of [`Button`](/library/api-reference/app-testing/testing-element-classes#sttestingv1element_treebutton). Aside from setting the value of a button with `.set_value()`, you can also use `.click()`. Check out each testing element class for its specific methods.
+
+### How to inspect elements
+
+All elements, including widgets, have a universal `.value` property. This returns the contents of the element. For widgets, this is same as the return value or value in Session State. Additionally, you can check many other details for widgets like labels or disabled status. Many parameters are available for inspection, but not all. Use linting software to see what is currently supported. Here's an example:
```

Comment:
Might be worth adding a sentence for what `.value` represents for non-widgets. Could be something like:

> For non-input elements, .value will be the primary contents argument - such as `body` for text or status elements like `st.markdown` or `st.error`, or `data` for data elements like `st.dataframe` or `st.json`.